### PR TITLE
feat(ui): paint and clear non-palette highlight colors

### PR DESCRIPTION
**Pull request by cameronapak**
## Review feedback addressed (Brenden / bmanquen)

**Spec / correctness**
- `buildVerseActionSwatches` now normalizes every `activeHighlights` entry once up front; remove/apply rows use that normalized set only (fixes uppercase/`#` palette colors missing from remove row and duplicate swatch keys).
- Tests: Story 10 (non-palette on part of selection still gets remove swatch); server `parseServerColors` path drops invalid hex.

**Docs**
- `docs/highlight-flow-statechart.md`: TAP_COLOR noop covers non-palette apply (YPE-4494).
- `docs/adr/YPE-3705-controlled-mode.md`: valid non-palette paint/clear in scope; custom **apply** creation still out of scope.

**Nits**
- Shared `stripHighlightHexPrefix` helper for validation + normalization.
- Comment on non-palette `.sort()` for stable tray order.
- `fillFor(hex)` in popover test instead of magic rgb string.

**Explicitly not done:** machine guard collapse, unified projector, export cleanup, `#` rejection tightening.

YPE-4494

<div><a href="https://cursor.com/agents/bc-206c027d-cc58-4a3c-bdde-56c747704e35?cursor_ref=pr_footer&cursor_cta=open_in_web"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-206c027d-cc58-4a3c-bdde-56c747704e35&cursor_ref=pr_footer&cursor_cta=open_in_cursor"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

[YPE-4494]: https://lifechurch.atlassian.net/browse/YPE-4494?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Visual context (YPE-4494)

Policy in one pass:

```text
on(API color / user action)
  invalid hex        -> drop (no paint, no swatch)
  valid non-palette  -> paint + exact Clear swatch; apply rejected
  palette            -> paint + Clear; apply allowed
```

Ownership:

```text
packages/ui/src/
|-- lib/
|   |-- highlight-colors.ts      # validate / normalize / swatch row
|   `-- highlight-projection.ts  # project any valid API hex into paint
|-- components/
|   |-- use-bible-reader-highlights.ts      # controlled + live projection
|   |-- bible-reader-highlights-machine.ts  # palette-only apply paths
|   `-- verse-action-popover.tsx            # Clear for exact color
```

Flow:

```mermaid
flowchart LR
  API[API / controlled highlight] --> Valid{Valid 6-digit hex?}
  Valid -->|no| Drop[Drop]
  Valid -->|yes| Paint[Normalize + paint]
  Paint --> Tray[Popover swatches]
  Tray --> Remove[Clear at exact hex]
  Apply[Apply tap] --> Gate{In HIGHLIGHT_COLORS?}
  Gate -->|yes| Write[Write highlight]
  Gate -->|no| Reject[Reject]
```

Pairs with expo [#131](https://github.com/youversion/platform-sdk-reactnative-expo/pull/131) for the native tray.

<h3>Greptile Summary</h3>

The PR extends Bible-reader highlight projection and removal to valid non-palette hexadecimal colors while retaining palette-only apply behavior and dropping invalid colors.
- Adds shared color validation, normalization, swatch construction, and highlight projection behavior.
- Updates controlled and self-contained highlight flows, verse presentation, tests, documentation, and the package changeset.

<h3>Confidence Score: 5/5</h3>

The PR appears safe to merge because no blocking failure remains.

No blocking failure remains.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/ui/src/lib/highlight-colors.ts | Centralizes hexadecimal validation and normalization and constructs apply/remove swatch rows with support for valid non-palette colors. |
| packages/ui/src/lib/highlight-projection.ts | Projects all valid normalized highlight colors into the rendered verse map while excluding invalid values. |
| packages/ui/src/components/use-bible-reader-highlights.ts | Integrates normalized server and controlled highlight projection with the existing adapter. |
| packages/ui/src/components/bible-reader-highlights-machine.ts | Enforces palette-only apply operations at the state-machine boundary while preserving remove behavior for projected colors. |
| packages/ui/src/components/verse-action-popover.tsx | Presents exact-color removal swatches for valid non-palette highlights. |
| packages/ui/src/components/verse.tsx | Updates verse highlight rendering and contrast behavior for arbitrary valid hexadecimal fills. |

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart LR
  A[API or controlled highlight] --> B{Valid six-digit hex?}
  B -->|No| C[Drop]
  B -->|Yes| D[Normalize and paint]
  D --> E[Expose exact-color clear swatch]
  F[User apply tap] --> G{Palette color?}
  G -->|Yes| H[Apply highlight]
  G -->|No| I[No-op]
```

<sub>Reviews (6): Last reviewed commit: ["docs(ui): note HIGHLIGHT\_COLORS re-expor..."](https://github.com/youversion/platform-sdk-react/commit/bce34ee9d226ee3527c5b60cf26f7f7f83fb7481) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=52633517)</sub>

**Context used:**

- Context used - AGENTS.md ([source](https://github.com/youversion/platform-sdk-react/blob/main/AGENTS.md))
- Knowledge Base — [packages/ui](https://app.greptile.com/youversion/-/custom-context/knowledge-base/youversion/platform-sdk-react/-/docs/ui-package.md)

### CODE REVIEW

File: packages/ui/src/lib/highlight-colors.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  selectedVerses: readonly number[];
+  highlightedVerses: Readonly<Record<number, string>>;
+};
+
+/**
+ * Builds the verse-action popover swatch row: remove circles (checkmark) for
+ * every distinct valid color on the selection — palette or not — then apply
+ * circles for palette colors only. Invalid hex never appears.
+ */
+export function buildVerseActionSwatches({
```
Comment (cameronapak) [inline packages/ui/src/lib/highlight-colors.ts:53]: **praise:** `buildVerseActionSwatches` matches the locked product rules — exact-hex remove (palette + non-palette), apply palette-only, invalid never appears.

*Code Reviewer bot, sent on behalf of Cam.*

File: packages/ui/src/lib/highlight-colors.test.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import { describe, expect, it } from 'vitest';
+import {
+  HIGHLIGHT_COLORS,
+  buildVerseActionSwatches,
+  isPaletteHighlightColor,
+  isValidHighlightHex,
+  normalizeHighlightHex,
+} from './highlight-colors';
+
+describe('isValidHighlightHex', () => {
```
Comment (cameronapak) [inline packages/ui/src/lib/highlight-colors.test.ts]: **issue:** Root `AGENTS.md` Musts: one optional top-level `describe` for the module; no nested `describe`. This file opens four sibling top-level `describe`s (`isValidHighlightHex`, `normalizeHighlightHex`, `isPaletteHighlightColor`, `buildVerseActionSwatches`).

Prefer one module `describe('highlight-colors')` (or flat `it`s) so the new file matches the Musts.

*Code Reviewer bot, sent on behalf of Cam.*

File: .changeset/ype-4494-non-palette-highlight-paint-clear.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+'@youversion/platform-core': patch
+'@youversion/platform-react-hooks': patch
```
Comment (cameronapak) [inline .changeset/ype-4494-non-palette-highlight-paint-clear.md]: **question:** This patches `platform-core` and `platform-react-hooks` with no code changes in those packages. Intentional for unified versioning, or should the changeset be `platform-react-ui` only?

*Code Reviewer bot, sent on behalf of Cam.*

**Review by cameronapak — COMMENTED** [review]
## Spec
**Pass** (minor soft edges) vs [YPE-4494](https://lifechurch.atlassian.net/browse/YPE-4494).

- Apply palette-only; non-palette apply → `noop`
- Paint accepts valid non-palette; invalid dropped
- `buildVerseActionSwatches`: exact-hex remove + checkmark; apply palette-only; ANY kept
- Hook/machine remove path allows non-palette clear
- Changeset present; no Swift/Kotlin

Soft: remove matching in places uses `toLowerCase()` rather than `normalizeHighlightHex` (`#` prefixed edge). Fine if API never sends `#`.

## Standards
**Mostly pass.**

1. **Hard (test shape):** new `highlight-colors.test.ts` has four top-level `describe`s. Root `AGENTS.md` Musts ask for one optional top-level `describe` per module (no nested). Flatten or use one module describe.
2. **Judgement:** changeset also patches `@youversion/platform-core` + `@youversion/platform-react-hooks` with no code changes there — confirm that’s intentional for unified versioning.
3. **Judgement:** palette-apply gates repeated across machine + hook (policy lives in `highlight-colors`, enforcement still sprinkled).

Product rules look right; docs/test-shape nits only.

---
*Code Reviewer bot, sent on behalf of Cam. Own-PR review → COMMENT (GitHub blocks self-approve).*

File: packages/ui/src/lib/highlight-colors.test.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import { describe, expect, it } from 'vitest';
+import {
+  HIGHLIGHT_COLORS,
+  buildVerseActionSwatches,
+  isPaletteHighlightColor,
+  isValidHighlightHex,
+  normalizeHighlightHex,
+} from './highlight-colors';
+
+describe('highlight-colors', () => {
```
Comment (cameronapak) [inline packages/ui/src/lib/highlight-colors.test.ts:10]: **praise:** Single module `describe('highlight-colors')` matches the AGENTS Musts. Thanks.

*Code Reviewer bot, sent on behalf of Cam.*

File: .changeset/ype-4494-non-palette-highlight-paint-clear.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+'@youversion/platform-react-ui': patch
```
Comment (cameronapak) [inline .changeset/ype-4494-non-palette-highlight-paint-clear.md:2]: **praise:** Scoped to `@youversion/platform-react-ui` only — clears the prior question.

*Code Reviewer bot, sent on behalf of Cam.*

**Review by cameronapak — COMMENTED** [review]
## Re-review @ `6cbfbbd`

Follow-up addresses the prior Standards notes.

## Spec
**Pass.** Unchanged product rules still match YPE-4494 (prior review).

## Standards
**Pass.**
- `highlight-colors.test.ts` now uses a single top-level `describe('highlight-colors')`
- Changeset is `platform-react-ui` only (core/hooks bump removed)

No remaining blocking findings from my side.

---
*Code Reviewer bot, sent on behalf of Cam. Own-PR review → COMMENT (GitHub blocks self-approve).*

File: .changeset/ype-4494-non-palette-highlight-paint-clear.md (hunk 3 above)
Comment (cameronapak) [reply .changeset/ype-4494-non-palette-highlight-paint-clear.md]: Already scoped to `@youversion/platform-react-ui` only. Resolving.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/lib/highlight-colors.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
+/**
```
Comment (bmanquen) [inline packages/ui/src/lib/highlight-colors.ts:1]: `praise:` one module now owns the palette/validity rules, and `deriveHighlightedVerses` losing its `allowedColors` param is a real simplification. Test file is flat with a single top-level `describe`, per root `AGENTS.md`.

File: packages/ui/src/lib/highlight-colors.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ * Builds the verse-action popover swatch row: remove circles (checkmark) for
+ * every distinct valid color on the selection — palette or not — then apply
+ * circles for palette colors only. Invalid hex never appears.
+ */
+export function buildVerseActionSwatches({
+  activeHighlights,
+  selectedVerses,
+  highlightedVerses,
+}: BuildVerseActionSwatchesInput): VerseActionSwatch[] {
+  const activePalette = HIGHLIGHT_COLORS.filter((color) => activeHighlights.has(color));
```
Comment (bmanquen) [inline packages/ui/src/lib/highlight-colors.ts]: `issue (non-blocking):` a non-lowercase palette color gets no remove swatch.

`activePalette` matches raw (`activeHighlights.has(color)`), while line 59 excludes anything `isPaletteHighlightColor` accepts — so `'FFFE00'` or `'#fffe00'` matches neither branch and disappears. Against story 7 ("existing palette apply and clear behavior to keep working") and story 3.

Same lines: `'AABBCC'` + `'aabbcc'` both survive → duplicate swatches and duplicate React `key`s.

Unreachable via `BibleReader` (it normalizes), but `VerseActionPopover` is public API taking `activeHighlights: Set<string>`. Normalizing once at the top of the builder closes both.

File: packages/ui/src/lib/highlight-colors.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+/** Canonical apply palette — yellow, green, blue, orange, pink (matches iOS). */
+export const HIGHLIGHT_COLORS = ['fffe00', '5dff79', '00d6ff', 'ffc66f', 'ff95ef'] as const;
+
+export type HighlightColor = (typeof HIGHLIGHT_COLORS)[number];
+
+const HIGHLIGHT_HEX_REGEX = /^[0-9a-f]{6}$/i;
+
+/** Whether `color` is a valid API highlight hex (6 digits, optional `#` stripped). */
+export function isValidHighlightHex(color: string): boolean {
```
Comment (bmanquen) [inline packages/ui/src/lib/highlight-colors.ts:18]: `issue (non-blocking):` Speculative Generality — `isValidHighlightHex` is exported but has no production caller (only `normalizeHighlightHex` and its own tests). Two of the seven `it`s exist solely for it; root `AGENTS.md` asks for the lightest falsifying flavor. Inline it and drop those cases.

File: packages/ui/src/lib/highlight-colors.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  const stripped = color.startsWith('#') ? color.slice(1) : color;
```
Comment (bmanquen) [inline packages/ui/src/lib/highlight-colors.ts]: `nitpick:` the `#` prefix widens validity past the spec, which pins the API contract to `/^[0-9a-f]{6}$/`. It's also the other half of the casing mismatch flagged at line 56. Duplicated Code too: this exact strip appears again in `normalizeHighlightHex`.

File: packages/ui/src/lib/highlight-colors.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+export function buildVerseActionSwatches({
+  activeHighlights,
+  selectedVerses,
+  highlightedVerses,
+}: BuildVerseActionSwatchesInput): VerseActionSwatch[] {
+  const activePalette = HIGHLIGHT_COLORS.filter((color) => activeHighlights.has(color));
+  const activeNonPalette = [...activeHighlights]
+    .map((color) => normalizeHighlightHex(color))
+    .filter((color): color is string => color !== null && !isPaletteHighlightColor(color))
+    .sort();
```
Comment (bmanquen) [inline packages/ui/src/lib/highlight-colors.ts:68]: `question:` why `.sort()` the non-palette remove swatches when palette ones keep canonical order? Determinism is plausible, but uncommented — the tray's ordering contract isn't self-documenting.

File: packages/ui/src/lib/highlight-colors.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+  const removeColors = [...activePalette, ...activeNonPalette];
+
+  const highlightedVerseCount = selectedVerses.filter((verse) => highlightedVerses[verse]).length;
+  const unHighlightedCount = selectedVerses.length - highlightedVerseCount;
+  const allPaletteColorsActive = HIGHLIGHT_COLORS.every((color) => activeHighlights.has(color));
```
Comment (bmanquen) [inline packages/ui/src/lib/highlight-colors.ts]: `praise:` swapping `size === 5` for `HIGHLIGHT_COLORS.every(...)` quietly fixes the apply-row miscount non-palette colors would have introduced.

File: packages/ui/src/components/bible-reader-highlights-machine.ts [hunk 12]
```diff
@@ -363,14 +364,20 @@ export const bibleReaderHighlightsMachine = setup({
     signedOut: ({ context }) => !context.isAuthenticated,
 
     // ── TAP_COLOR fork ──
-    tapInert: ({ event }) => event.type === 'TAP_COLOR' && event.verses.length === 0,
+    tapInert: ({ event }) =>
```
Comment (bmanquen) [inline packages/ui/src/components/bible-reader-highlights-machine.ts:367]: `thought:` `isPaletteHighlightColor(event.color)` now guards six sites (`tapInert`, `tapCanWrite`, `tapNeedsSignIn`, `startApplyWrite:481`, `stashPendingTap:508`, resume loop `:554`). One normalized-and-validated tap value would collapse them.

Spec only asks to "Reject non-palette colors on apply writes" — the transition guards already do that, so the three action-level repeats are belt-and-braces.

File: packages/ui/src/lib/highlight-projection.ts [hunk 13]
```diff
@@ -50,24 +51,21 @@ export function expandPassageId(passageId: string): ExpandedPassageId | null {
...
   book: string,
   chapter: string,
-  allowedColors?: readonly string[],
 ): Record<number, string> {
   const map: Record<number, string> = {};
   for (const { version_id, passage_id, color } of highlights) {
     if (version_id !== versionId) continue;
-    const normalizedColor = color.toLowerCase();
-    if (allowedColors && !allowedColors.includes(normalizedColor)) continue;
+    const normalizedColor = normalizeHighlightHex(color);
```
Comment (bmanquen) [inline packages/ui/src/lib/highlight-projection.ts:67]: `thought:` this normalize-then-skip shape is now duplicated in `use-bible-reader-highlights.ts:114` (`parseServerColors`) — same rule, two projections.

File: packages/ui/src/components/use-bible-reader-highlights.ts [hunk 14]
```diff
@@ -111,8 +111,10 @@ function parseServerColors(
   for (const highlight of highlights?.data ?? []) {
     if (highlight.version_id !== versionId) continue;
     if (!highlight.passage_id.startsWith(versePrefix)) continue;
+    const normalizedColor = normalizeHighlightHex(highlight.color);
+    if (normalizedColor === null) continue;
```
Comment (bmanquen) [inline packages/ui/src/components/use-bible-reader-highlights.ts:115]: `suggestion (non-blocking):` locked seam 2's "invalid hex is dropped" is only proven on the controlled projection. `parseServerColors` gained the same drop but has no test — the new server-path test covers only valid non-palette paint.

File: packages/ui/src/lib/highlight-colors.test.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+      selectedVerses: [1, 2],
+      highlightedVerses: { 1: 'not-valid', 2: HIGHLIGHT_COLORS[0] },
+    });
+
+    expect(swatches.filter((swatch) => swatch.showRemove)).toEqual([
+      { color: HIGHLIGHT_COLORS[0], showRemove: true, key: `${HIGHLIGHT_COLORS[0]}-clear` },
+    ]);
+  });
+
+  it('buildVerseActionSwatches shows remove for every distinct valid color on a mixed selection (ANY rule)', () => {
```
Comment (bmanquen) [inline packages/ui/src/lib/highlight-colors.test.ts:78]: `suggestion (non-blocking):` story 10 ("a non-palette color that covers only part of my selection to still get a remove swatch") has no case. This ANY-rule test uses a fully-covered two-color selection; nothing mixes a non-palette verse with an unhighlighted one. Correct by construction, but this is the locked seam where the spec wants it pinned.

File: packages/ui/src/components/verse-action-popover.test.tsx [hunk 16]
```diff
@@ -411,6 +411,22 @@ describe('VerseActionPopover', () => {
...
+      <VerseActionPopover
+        {...defaultProps}
+        activeHighlights={new Set([custom])}
+        selectedVerses={[1]}
+        highlightedVerses={{ 1: custom }}
+      />,
+    );
+
+    const removeButtons = clearButtons();
+    expect(removeButtons).toHaveLength(1);
```
Comment (bmanquen) [inline packages/ui/src/components/verse-action-popover.test.tsx]: `nitpick:` magic `'rgb(170, 187, 204)'` and a theme-coupled style assertion. `bible-reader-controlled.test.tsx` already models the `fillFor(hex)` helper idiom.

**Review by bmanquen — COMMENTED** [review]
Two-axis review — **Standards** (repo conventions + Fowler smell baseline) and **Spec** (YPE-4494, web half only; RN Expo lands on `platform-sdk-reactnative-expo#100`). Findings are inline, as [conventional comments](https://conventionalcomments.org/). Nothing fixed — reporting only.

No hard standards violations. All web-half spec requirements are met: apply stays palette-only, paint accepts any valid hex, remove swatches use exact hex + checkmark, invalid hex dropped from paint and tray, ANY rule unchanged. 129 tests pass.

One finding has no diff line to anchor to:

`suggestion (non-blocking): docs/highlight-flow-statechart.md:29` — "TAP_COLOR forks in `flow`: authorized / signed out / no permission" no longer describes the machine; `tapInert` now also swallows non-palette colors. `bible-reader.tsx:241` prose was updated, the statechart and `docs/adr/YPE-3705-controlled-mode.md` ("Out of scope: … custom colors") were not.

File: packages/ui/src/components/verse-action-popover.tsx [hunk 17]
```diff
@@ -6,19 +6,12 @@ import { cn } from '../lib/utils';
 import { BoxStackIcon } from './icons/box-stack';
 import { BoxArrowUpIcon } from './icons/box-arrow-up';
 import { CheckIcon } from './icons/check';
-import { hexToRgba, HIGHLIGHT_FILL_OPACITY_DARK } from './verse';
+import { buildVerseActionSwatches } from '@/lib/highlight-colors';
```
Comment (bmanquen) [inline packages/ui/src/components/verse-action-popover.tsx:9]: `nitpick:` mixed import specifiers in one file — `@/lib/highlight-colors` next to `'../lib/utils'`.

File: packages/ui/src/components/verse-action-popover.tsx [hunk 18]
```diff
@@ -6,19 +6,12 @@ import { cn } from '../lib/utils';
...
-type Measurable = { getBoundingClientRect: () => DOMRect };
-
-/**
- * Highlight colors, as 6-digit lowercase hex (no `#`) so they map 1:1 onto the
- * API `highlight.color` field (/^[0-9a-f]{6}$/). Order is the canonical apply
- * order: yellow, green, blue, orange, pink. Hardcoded to match the YouVersion
- * iOS app exactly.
- */
-export const HIGHLIGHT_COLORS = ['fffe00', '5dff79', '00d6ff', 'ffc66f', 'ff95ef'] as const;
+export { HIGHLIGHT_COLORS, type HighlightColor } from '@/lib/highlight-colors';
```
Comment (bmanquen) [inline packages/ui/src/components/verse-action-popover.tsx:13]: `nitpick:` this re-export is a Middle Man kept only for back-compat — worth a comment saying so, or a deprecation path.

File: packages/ui/src/lib/highlight-colors.ts (hunk 9 above)
Comment (cameronapak) [reply packages/ui/src/lib/highlight-colors.ts]: Shared `stripHighlightHexPrefix` now. Keeping optional `#` strip: public `activeHighlights: Set<string>` can arrive CSS-ish; API persistence still lands as 6-digit lowercase via `normalizeHighlightHex`. Resolving.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/lib/highlight-colors.ts (hunk 10 above)
Comment (cameronapak) [reply packages/ui/src/lib/highlight-colors.ts:68]: Comment is on the `.sort()` now: deterministic tray order across renders; palette stays canonical above. Resolving.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/components/use-bible-reader-highlights.ts (hunk 14 above)
Comment (cameronapak) [reply packages/ui/src/components/use-bible-reader-highlights.ts:115]: Pinned in `862ecde` — `drops invalid hex from the fetched server path (parseServerColors)`. Resolving.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/lib/highlight-colors.test.ts (hunk 15 above)
Comment (cameronapak) [reply packages/ui/src/lib/highlight-colors.test.ts:78]: Pinned in `862ecde` — `shows a remove swatch when non-palette covers only part of the selection (Story 10)`. Resolving.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/lib/highlight-colors.ts (hunk 8 above)
Comment (cameronapak) [reply packages/ui/src/lib/highlight-colors.ts:18]: Keeping `isValidHighlightHex` as the named validity seam (same helper on the RN Expo side for YPE-4494). Production still goes through `normalizeHighlightHex`; the dedicated `it`s are the lightest pin of the regex contract itself. Not inlining this pass.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/components/bible-reader-highlights-machine.ts (hunk 12 above)
Comment (cameronapak) [reply packages/ui/src/components/bible-reader-highlights-machine.ts:367]: Agreed it’s belt-and-braces past the transition guards. Skipping the collapse in this PR — would widen the machine diff without a behavior change.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/lib/highlight-projection.ts (hunk 13 above)
Comment (cameronapak) [reply packages/ui/src/lib/highlight-projection.ts:67]: Same rule, two call sites — leaving both for now. A shared projector would be a follow-up, not needed to ship YPE-4494.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/components/verse-action-popover.test.tsx (hunk 16 above)
Comment (cameronapak) [reply packages/ui/src/components/verse-action-popover.test.tsx]: Skipping `fillFor` extract here — style-only, and this file already has a lot of pre-existing assertion shapes.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/components/verse-action-popover.tsx (hunk 17 above)
Comment (cameronapak) [reply packages/ui/src/components/verse-action-popover.tsx:9]: Skipping — cosmetic; file already mixes `@/` and relative elsewhere. Not worth churn this PR.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/components/verse-action-popover.tsx (hunk 18 above)
Comment (cameronapak) [reply packages/ui/src/components/verse-action-popover.tsx:13]: Skipping a deprecation path this PR — back-compat re-export stays. Happy to add a one-line comment in a tiny follow-up if you want it explicit.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/lib/highlight-colors.ts (hunk 7 above)
Comment (cameronapak) [reply packages/ui/src/lib/highlight-colors.ts]: Fixed in `862ecde` — `buildVerseActionSwatches` normalizes once into `normalizedActive` (covers uppercase / `#` palette + dedupes casings). Tests pin both. Resolving.

*Code Implementor bot, sent on behalf of Cam. I am an AI.*

File: packages/ui/src/components/verse-action-popover.test.tsx [hunk 19]
```diff
@@ -411,7 +418,42 @@ describe('VerseActionPopover', () => {
...
+    expect(removeButtons[0]!.style.backgroundColor).toBe('rgb(170, 187, 204)');
+  });
+
   describe('Edge cases', () => {
+<<<<<<< HEAD
+=======
```
Comment (bmanquen) [inline packages/ui/src/components/verse-action-popover.test.tsx:439]: Looks like some conflicts left.

File: packages/ui/src/components/verse-action-popover.test.tsx [hunk 20]
```diff
@@ -411,7 +418,42 @@ describe('VerseActionPopover', () => {
...
+          highlightedVerses={{ 1: custom }}
+        />,
+      );
+
+      const removeButtons = clearButtons();
+      expect(removeButtons).toHaveLength(1);
+      expect(removeButtons[0]!.style.backgroundColor).toBe(fillFor(custom));
+    });
+
+>>>>>>> 16594fe (fix(ui): address Brenden review on non-palette swatch normalization (YPE-4494))
```
Comment (bmanquen) [inline packages/ui/src/components/verse-action-popover.test.tsx:456]: blocking: merge conflict left in.

**Review by bmanquen — CHANGES_REQUESTED** [review]

**Review by bmanquen — APPROVED** [review]
